src/Agent.Worker/DiagnosticLogManager.cs

```diff
@@ -131,6 +139,46 @@ public sealed class DiagnosticLogManager : AgentService, IDiagnosticLogManager
             executionContext.Debug("Diagnostic file upload complete.");
         }
 
+        private async Task DumpCloudInitLogs(DateTime jobStartTimeUtc, string diagFolder)
+        {
+            var builder = new StringBuilder();
+
+            string cloudInit = WhichUtil.Which("cloud-init", trace: Trace);
+            if (string.IsNullOrEmpty(cloudInit))
+            {
+                Trace.Info("Cloud-init ins't found on current machine.");
+                return;
+            }
+            
+            string resultName = $"cloudinit-{jobStartTimeUtc.ToString("yyyyMMdd-HHmmss")}-logs.tar.gz";
+            string arguments = $"collect-logs -t \"{diagFolder}/{resultName}\"";
+
+            using (var processInvoker = HostContext.CreateService<IProcessInvoker>())
+            {
+                processInvoker.OutputDataReceived += (object sender, ProcessDataReceivedEventArgs args) =>
+                {
+                    builder.AppendLine(args.Data);
+                };
+
+                processInvoker.ErrorDataReceived += (object sender, ProcessDataReceivedEventArgs args) =>
+                {
+                    builder.AppendLine(args.Data);
+                };
+
+                await processInvoker.ExecuteAsync(
+                    workingDirectory: HostContext.GetDirectory(WellKnownDirectory.Bin),
+                    fileName: cloudInit,
+                    arguments: arguments,
+                    environment: null,
+                    requireExitCodeZero: false,
+                    outputEncoding: null,
+                    killProcessOnCancel: false,
+                    cancellationToken: default(CancellationToken));
+            }
+
+            Trace.Info(builder.ToString());
```

Comment:
Please split the logic of this method up - this method searches for cloud-init, calls logs collection and trace retrieved data - could it make sense at least to move tracing out of this method?